### PR TITLE
Lecture 2

### PR DESCRIPTION
**Pull request by gznznzjsn**

### CODE REVIEW

File: src/main/java/com/es/phoneshop/dao/impl/ArrayListProductDao.java [hunk 1]
```diff
@@ -4,23 +4,30 @@
...
+import java.util.Comparator;
 import java.util.List;
 import java.util.concurrent.locks.ReadWriteLock;
 import java.util.concurrent.locks.ReentrantReadWriteLock;
 import java.util.stream.Collectors;
 
 public class ArrayListProductDao implements ProductDao {
+    private static ProductDao instance;
+
+    public static synchronized ProductDao getInstance() {
```
Comment (nd524s) [inline src/main/java/com/es/phoneshop/dao/impl/ArrayListProductDao.java]: No need to synchronize on method level. It brakes all the advantages of this kind of singletones

File: src/main/java/com/es/phoneshop/dao/impl/SortField.java [hunk 2]
```diff
@@ -0,0 +1,5 @@
+package com.es.phoneshop.dao.impl;
+
+public enum SortField {
+    description, price
```
Comment (nd524s) [inline src/main/java/com/es/phoneshop/dao/impl/SortField.java]: Use upper case constants in enum

File: src/main/java/com/es/phoneshop/dao/impl/SortOrder.java [hunk 3]
```diff
@@ -0,0 +1,5 @@
+package com.es.phoneshop.dao.impl;
+
+public enum SortOrder {
+    asc, desc
```
Comment (nd524s) [inline src/main/java/com/es/phoneshop/dao/impl/SortOrder.java]: Use upper case constants in enum

Comment (nd524s) [inline src/main/java/com/es/phoneshop/dao/impl/SortOrder.java]: Create package called enum on dao package level and move enums there

File: src/main/java/com/es/phoneshop/model/PriceHistoryBin.java [hunk 4]
```diff
@@ -0,0 +1,28 @@
+package com.es.phoneshop.model;
+
+import java.math.BigDecimal;
+import java.time.LocalDate;
+
+public class PriceHistoryBin {
```
Comment (nd524s) [inline src/main/java/com/es/phoneshop/model/PriceHistoryBin.java:6]: Rename to PriceHistoryBean

File: src/main/java/com/es/phoneshop/web/DemoDataServletContextListener.java [hunk 5]
```diff
@@ -0,0 +1,56 @@
+package com.es.phoneshop.web;
```
Comment (nd524s) [inline src/main/java/com/es/phoneshop/web/DemoDataServletContextListener.java]: You can create listener package and move this class to it

File: src/main/java/com/es/phoneshop/web/ProductListPageServlet.java [hunk 6]
```diff
@@ -2,26 +2,36 @@
...
     }
 
     @Override
     protected void doGet(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
-        request.setAttribute("products", productDao.findProducts());
+        String query = request.getParameter("query");
+        String sortField = request.getParameter("sort");
+        String sortOrder = request.getParameter("order");
+        request.setAttribute("products",
+                productDao.findProducts(query,
```
Comment (nd524s) [inline src/main/java/com/es/phoneshop/web/ProductListPageServlet.java:31]: When I click on sorting without queries I see this URL http://localhost:8080/phoneshop-servlet-api/products?sort=description&order=asc&query=
Remove query parameter if it doesn't have any value

File: src/main/webapp/WEB-INF/tags/master.tag [hunk 7]
```diff
@@ -17,5 +18,11 @@
   <main>
     <jsp:doBody/>
   </main>
+  <footer>
```
Comment (nd524s) [inline src/main/webapp/WEB-INF/tags/master.tag]: You can move header and footer to separate components

**Review by nd524s — CHANGES_REQUESTED** [review]

File: src/main/java/com/es/phoneshop/model/PriceHistoryBin.java (hunk 4 above)
Comment (gznznzjsn) [reply src/main/java/com/es/phoneshop/model/PriceHistoryBin.java:6]: Под bin  я имею в виду бинарность класса, что он используется как пара

**Review by nd524s — APPROVED** [review]

**Pull request merged**